feat(nydusify): support zero-disk transfer

daiyongxuan opened the pull request.

imeoer commented on contrib/nydusify/pkg/converter/provider/stream_content.go:
```diff
@@ -0,0 +1,226 @@
+	s.mu.RLock()
+	ref := s.defaultRef
+	s.mu.RUnlock()
+	if ref == "" {
```
May be:

```
s.mu.RUnlock()
s.mu.RLock()
ref := s.defaultRef
s.mu.RUnlock()
```

to

```
ref := s.defaultRef
s.mu.RUnlock()
```

imeoer commented on contrib/nydusify/pkg/converter/provider/stream_content.go:
```diff
@@ -0,0 +1,226 @@
+	return ctrcontent.Status{}, errdefs.ErrNotFound
+}
+
+func (s *StreamContent) ListStatuses(ctx context.Context, filters ...string) ([]ctrcontent.Status, error) {
```
The lint checks should be fixed. :)

imeoer commented on contrib/nydusify/pkg/converter/provider/stream_content.go:
```diff
@@ -0,0 +1,226 @@
+	if int64(w.buf.Len()) > size {
+		b := w.buf.Bytes()
+		w.buf.Reset()
+		_, _ = w.buf.Write(b[:size])
```
How about `w.buf.Truncate(size)` ?

imeoer commented on contrib/nydusify/pkg/converter/provider/stream_content.go:
```diff
@@ -0,0 +1,226 @@
+// Ingester
+func (s *StreamContent) Writer(_ context.Context, opts ...ctrcontent.WriterOpt) (ctrcontent.Writer, error) {
+	// Heuristic routing:
+	// - If ref looks like containerd fetch key (manifest-*/index-*/layer-*/config-*/attestation-*),
```
Can we treat all fetch key as remote content here?

daiyongxuan committed:
feat: support customized provider with stream content store

This commit introduces support for customized content providers in nydusify,
with a new StreamContent implementation that enables:

- Stream-based content storage without local ingestion
- Direct remote reading using registry HTTP range requests
- In-memory label management for handler pipeline compatibility
- Optimized copying workflow for large container images

Changes:
- Add optional storeOverride parameter to provider.New()
- Implement StreamContent as content.Store adapter
- Update all provider.New() calls to support new parameter
- Add stream-based copying in copier package
- Enhance ported.go with better remote handling

The StreamContent store provides a no-local-ingestion approach that
significantly reduces disk usage and improves performance for
container image processing workflows.

Signed-off-by: daiyongxuan <daiyongxuan20@mails.ucas.ac.cn>

imeoer reviewed (approved):
LGTM, thanks!
